File: backend/app/realtime/languages.py
"""Realtime voice routing — which app languages Qwen realtime S2S supports.

v11 M1 (2026-08-08). Source of truth for the realtime/cascade split; the
/api/languages endpoint exposes this as a per-language `realtime` flag so
the frontend routes without duplicating the list.

Verified against the official qwen3.5-omni realtime voice table
(https://docs.qwencloud.com/developer-guides/speech/realtime-multimodal-speech,
fetched 2026-08-08): every preset voice lists the same 29 output languages —
Chinese (incl. Cantonese), English, French, German, Russian, Italian,
Spanish, Portuguese, Japanese, Korean, Thai, Indonesian, Arabic, Vietnamese,
Turkish, Finnish, Polish, Hindi, Dutch, Czech, Urdu, Tagalog, Swedish,
Danish, Hebrew, Icelandic, Malay, Norwegian, Persian.

Mapped onto the app's 31 codes: zh-TW runs on Chinese voices (Taiwanese-
accented presets exist), fil runs on Tagalog, yue is native. Greek,
Bengali, Swahili, Azerbaijani and Tamil have no realtime voice → those
sessions stay on the cascade (typed/recorded) engine.
"""

REALTIME_LANGS: frozenset[str] = frozenset({
    "en", "es", "fr", "de", "it", "pt", "ru", "ja", "ko",
    "zh", "zh-TW", "yue",
    "ar", "hi", "th", "vi", "id", "tr", "nl", "pl", "sv",
    "he", "ur", "fil", "ms", "cs",
})

# App codes with no Qwen realtime voice — the cascade path keeps serving
# them. Documented for the routing decision, not enforced here.
CASCADE_ONLY_LANGS: frozenset[str] = frozenset({"el", "bn", "sw", "az", "ta"})


def supports_realtime(code: str) -> bool:
    """True when the language can run a realtime speech-to-speech session."""
    return code in REALTIME_LANGS

File: frontend/src/i18n/languages.js
// All supported languages, sorted alphabetically by English name
const LANGUAGES = [
  { code: 'ar', native: 'العربية', english: 'Arabic' },
  { code: 'az', native: 'Azərbaycan dili', english: 'Azerbaijani' },
  { code: 'bn', native: 'বাংলা', english: 'Bengali' },
  { code: 'yue', native: '粵語', english: 'Cantonese' },
  { code: 'cs', native: 'Čeština', english: 'Czech' },
  { code: 'nl', native: 'Nederlands', english: 'Dutch' },
  { code: 'en', native: 'English', english: 'English' },
  { code: 'fil', native: 'Filipino', english: 'Filipino' },
  { code: 'fr', native: 'Français', english: 'French' },
  { code: 'de', native: 'Deutsch', english: 'German' },
  { code: 'el', native: 'Ελληνικά', english: 'Greek' },
  { code: 'he', native: 'עברית', english: 'Hebrew' },
  { code: 'hi', native: 'हिन्दी', english: 'Hindi' },
  { code: 'id', native: 'Bahasa Indonesia', english: 'Indonesian' },
  { code: 'it', native: 'Italiano', english: 'Italian' },
  { code: 'ja', native: '日本語', english: 'Japanese' },
  { code: 'ko', native: '한국어', english: 'Korean' },
  { code: 'ms', native: 'Bahasa Melayu', english: 'Malay' },
  { code: 'zh', native: '中文', english: 'Mandarin Chinese' },
  { code: 'pl', native: 'Polski', english: 'Polish' },
  { code: 'pt', native: 'Português', english: 'Portuguese' },
  { code: 'ru', native: 'Русский', english: 'Russian' },
  { code: 'es', native: 'Español', english: 'Spanish' },
  { code: 'sw', native: 'Kiswahili', english: 'Swahili' },
  { code: 'sv', native: 'Svenska', english: 'Swedish' },
  { code: 'ta', native: 'தமிழ்', english: 'Tamil' },
  { code: 'th', native: 'ไทย', english: 'Thai' },
  { code: 'zh-TW', native: '繁體中文', english: 'Traditional Chinese' },
  { code: 'tr', native: 'Türkçe', english: 'Turkish' },
  { code: 'ur', native: 'اردو', english: 'Urdu' },
  { code: 'vi', native: 'Tiếng Việt', english: 'Vietnamese' },
];

const FLAGS = {
  ar: '🇸🇦', az: '🇦🇿', bn: '🇧🇩', yue: '🇭🇰', cs: '🇨🇿', nl: '🇳🇱',
  en: '🇬🇧', fil: '🇵🇭', fr: '🇫🇷', de: '🇩🇪', el: '🇬🇷', he: '🇮🇱',
  hi: '🇮🇳', id: '🇮🇩', it: '🇮🇹', ja: '🇯🇵', ko: '🇰🇷', ms: '🇲🇾',
  zh: '🇨🇳', 'zh-TW': '🇹🇼', pl: '🇵🇱', pt: '🇵🇹', ru: '🇷🇺', es: '🇪🇸',
  sw: '🇰🇪', sv: '🇸🇪', ta: '🇮🇳', th: '🇹🇭', tr: '🇹🇷', ur: '🇵🇰',
  vi: '🇻🇳',
};

export function flagFor(code) {
  return FLAGS[code] || '';
}

export default LANGUAGES;
